jvm-libs/linea/blob-decompressor/src/main/kotlin/net/consensys/linea/blob/GoNativeBlobDecompressor.kt

```diff
@@ -0,0 +1,105 @@
+  fun decompress(blob: ByteArray): ByteArray
+}
+
+internal class Adapter(
+  private val delegate: GoNativeBlobDecompressorJnaBinding,
+  private val maxExpectedCompressionRatio: Int = 10,
+  dictionaries: List<Path>
+) : BlobDecompressor {
+  init {
+    delegate.Init()
+
+    val paths = dictionaries.joinToString(separator = ":") { path -> path.toString() }
+
+    if (delegate.LoadDictionaries(paths) != dictionaries.size) {
+      throw DecompressionException("Failed to load dictionaries '$paths', error='${delegate.Error()}'")
+    }
+  }
+
+  override fun decompress(blob: ByteArray): ByteArray {
+    val decompressionBuffer = ByteArray(blob.size * maxExpectedCompressionRatio)
+    val decompressedSize = delegate.Decompress(blob, blob.size, decompressionBuffer, decompressionBuffer.size)
+    if (decompressedSize < 0) {
+      throw DecompressionException("Decompression failed, error='${delegate.Error()}'")
+    }
+    return decompressionBuffer.copyOf(decompressedSize)
+  }
+}
+
+internal interface GoNativeBlobDecompressorJnaBinding {
+
+  /**
+   * Init initializes the Decompressor. Must be run before anything else.
+   */
+  fun Init()
+
+  /**
+   * LoadDictionaries attempts to cache dictionaries from given paths, separated by colons,
+   * e.g. "../compressor_dict.bin:./other_dict"
+   * Returns the number of dictionaries successfully loaded, and -1 in case of failure, in which case Error() will
+   * return a description of the error.
+   *
+   * @param dictPaths a colon-separated list of paths to dictionaries, to be loaded into the decompressor
+   * @return the number of dictionaries loaded if successful, -1 if not.
+   */
+  fun LoadDictionaries(dictPaths: String): Int
+
+  /**
+
+   * Decompress processes a Linea blob and outputs an RLP encoded list of blocks.
+   * Due to information loss during pre-compression encoding, two pieces of information are represented "hackily":
+   * The block hash is in the ParentHash field.
+   * The transaction from address is in the signature.R field.
+   *
+   * Returns the number of bytes in out, or -1 in case of failure
+   * If -1 is returned, the Error() method will return a string describing the error.
+   *
+   * @param blob to be decompressed
+   * @param blob_len length of the blob
+   * @param out buffer to write the decompressed data
+   * @param out_max_len maximum length of the out buffer
+   * @return number of bytes in out, or -1 in case of failure
+   */
+  fun Decompress(blob: ByteArray, blob_len: Int, out: ByteArray, out_max_len: Int): Int
+
+  /**
+   * Error returns the last error message. Should be checked if Write returns false.
+   */
+  fun Error(): String?
+}
+
+internal interface GoNativeBlobDecompressorJnaLib : GoNativeBlobDecompressorJnaBinding, Library
+
+enum class BlobDecompressorVersion(val version: String) {
+  V1_1_0("v1.1.0")
+}
+
+class GoNativeBlobDecompressorFactory {
+  companion object {
+    private const val DICTIONARY_NAME = "compressor_dict.bin"
```

Comment:
Is this "compressor_dict.bin" required to be the same as the one used by `BlobCompressor`? If so, should we make sure they're referring to the same file instead of having two same copies for `BlobCompressor` and `BlobDecompressor` in the repo?

Reply:
It is for now, by coincidence. 
However, if tomorrow we have a new compressor with a new dictionary, then the decompressor will need to have both files because it needs to be backward compatible.